Ropchain constraint syntax parser

giridharprasath opened the pull request:
- Parse the rop constraints
- Update esil output to rzil sdb
- Add /Rg gadget command

 <!-- Filling this template is mandatory -->

Current ROP chain parser misses some gadget when compared with the ROPGadget
Asm:
```
section .text
global _start
_start:
mov dword [esp+0x0],0x80484b0
ret
```
ROPgadget output:
![ScreenShot_2024-06-13_at_06:15:20-PM](https://github.com/rizinorg/rizin/assets/17312540/9f1db093-ce80-47ff-affe-720647e83226)

Rizin output before
![ScreenShot_2024-06-13_at_06:14:47-PM](https://github.com/rizinorg/rizin/assets/17312540/3dd72547-6761-4ba2-a256-057921aea451)

Rizin output now:
![ScreenShot_2024-06-13_at_06:15:54-PM](https://github.com/rizinorg/rizin/assets/17312540/bcff7090-b37b-423a-b6ce-fc5f9f4cc990)

These ROP gadgets are needed for crafting better rop chains.

Partially fix: #2080 

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

Commit:
Ropchain constraint syntax parser

- Parse the rop constraints
- Update esil output to rzil sdb
- Add /Rg gadget command

XVilka commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -727,3 +730,389 @@ static void rop_classify(RzCore *core, Sdb *db, RzList /*<char *>*/ *ropList, co
 	free(str);
 }
+
+RZ_API void rz_rop_constraint_free(void *data) {
```
What is not related to the command handling, I recommend moving to a separate file in `librz/core`. E.g. `librz/core/crop.c` or something like that, where the main logic could be moved to.

XVilka replied:
For all `*_free()` functions we usually allow the argument to be NULL and add a notation `RZ_NULLABLE` to that argument.

XVilka commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -727,3 +730,389 @@ static void rop_classify(RzCore *core, Sdb *db, RzList /*<char *>*/ *ropList, co
+
+RZ_API void rz_rop_constraint_free(void *data) {
+	RzRopConstraint *constraint = data;
+	if (constraint) {
```
In such cases you could use the "return early" pattern and reverse the condition, it will make the code slightly easier to read.

XVilka commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -727,3 +730,389 @@ static void rop_classify(RzCore *core, Sdb *db, RzList /*<char *>*/ *ropList, co
+	return true;
+}
+
+static bool parse_il_op(RzList *args, const char *str, int *idx) {
```
This kind of code is better to be implemented in a general way and moved into `librz/il`, also covered with tests. @wargio @thestr4ng3r @DMaroo @Rot127  any thoughts?

XVilka replied:
Or into `librz/core/cil.c` where printing is already implemented.

giridharprasath replied:
I have already started writing unit tc for these APIs

DMaroo replied:
I don't get why we are parsing the IL as a string. Why can't we just pass an `RzILOpEffect *` for every instruction instead of parsing the stringified IL?

giridharprasath replied:
This is not completely parsed to IL yet. I have just stringified `RzILInstructionType`, the plan is to store the registers and op as `RzILOpEffect` as you have mentioned.

DMaroo replied:
No, I would like if there wasn't any parsing at all. Rizin is a single application and it should be able to pass around data without needing to stringify it and then (re)parse it. I must be missing something, because I don't get why any sort of parsing is necessary.

Rot127 replied:
I agree with @XVilka. Better move this logic to `librz/il`.

@DMaroo The ROP gadget constraints are given as string by the user. And need to be translated to their RzIL representation for matching. @giridharprasath This was the design you had in mind, correct?

giridharprasath replied:
Yes, that's right 👍

Commit:
Refactor cmd_search.c and add unit tc

wargio commented on librz/core/rop.c:
```diff
@@ -0,0 +1,774 @@
+		RzAsmOp *asmop = rz_asm_op_new();
+		switch (state->mode) {
+		case RZ_OUTPUT_MODE_JSON:
+			buf = malloc(hit->len);
+			if (!buf) {
+				goto cleanup;
+			}
+			rz_io_read_at(core->io, hit->addr, buf, hit->len);
+			rz_asm_set_pc(core->rasm, hit->addr);
+			rz_asm_disassemble(core->rasm, asmop, buf, hit->len);
+			rz_analysis_op_init(&aop);
+			rz_analysis_op(core->analysis, &aop, hit->addr, buf, hit->len, RZ_ANALYSIS_OP_MASK_ESIL);
+			size += hit->len;
+			if (aop.type != RZ_ANALYSIS_OP_TYPE_RET) {
+				char *opstr_n = rz_str_newf(" %s", RZ_STRBUF_SAFEGET(&aop.esil));
+				rz_list_append(ropList, opstr_n);
+			}
+			pj_o(state->d.pj);
+			pj_kn(state->d.pj, "offset", hit->addr);
+			pj_ki(state->d.pj, "size", hit->len);
+			pj_ks(state->d.pj, "opcode", rz_asm_op_get_asm(asmop));
+			pj_ks(state->d.pj, "type", rz_analysis_optype_to_string(aop.type));
+			pj_end(state->d.pj);
+			free(buf);
+			rz_analysis_op_fini(&aop);
```
this is a lot of code within the switch. probably is better to make a static function which handles only this form of printing.

wargio commented on librz/core/rop.c:
```diff
@@ -0,0 +1,774 @@
+			break;
+		case RZ_OUTPUT_MODE_QUIET:
+			// Print gadgets in a 'linear manner', each sequence on one line.
+			buf = malloc(hit->len);
+			if (!buf) {
+				goto cleanup;
+			}
```
this can be defined before the switch to avoid repetition and bugs

wargio commented on librz/core/rop.c:
```diff
@@ -0,0 +1,774 @@
+				continue;
+			}
+			if (is_end_gadget(&end_gadget, crop)) {
+#if 0
+				if (search->maxhits && rz_list_length (end_list) >= search->maxhits) {
+					// limit number of high level rop gadget results
+					rz_analysis_op_fini (&end_gadget);
+					break;
+				}
+#endif
```
if this is not used, you can remove it

Commit:
Initial commit for RzILOpEffect rop APIs

- This only adds the list of modified registers to a Structure

- TODO: move the APis to a separate file under il/ and add support for all the Effect and Pure

Commit:
WIP: Add RzILOpEffect ROP gadget analysis APIs

- Refactor code in rop.c
- fix memory leaks
- Add APIs for storing the information of a given gadget
- fix tc

Commit:
tiny rop.c cleanup

Rot127 commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,9 +4,11 @@
 #include "rz_types_base.h"
+#include "rz_rop.h"
+
+#include <subprojects/capstone-next/include/capstone/x86.h>
```
```suggestion
#include <capstone/x86.h>
```

Rot127 commented on librz/il/meson.build:
```diff
@@ -21,6 +21,7 @@ rz_il_sources = [
    'il_validate.c',
    'il_vm.c',
    'il_vm_eval.c',
+    'il_rop_gadget.c',
```
```suggestion
   'il_rop_gadget.c',
```

Rot127 commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,70 @@
+
+typedef enum rzil_instr_type {
+	// Register to register
+	MOV_CONST, // reg <- const
```
```suggestion
	MOV_CONST, ///< reg <- const
```

Also below

Rot127 commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2020-2021 z3phyr <giridh1337@gmail.com>
```
```suggestion
// SPDX-FileCopyrightText: 2024 z3phyr <giridh1337@gmail.com>
```

Commit:
WIP: Add rop gadget semantics

- RzIL events parsing
- Store register info

giridharprasath commented:
```
  0x00400565               0000  add byte [rax], al
  0x00400567           007cffff  add byte [rdi + rdi*8 - 1], bh
  0x0040056b           ff6c0000  jmp [rax + rax]
Gadget size: 10

Current /Rg output:
Gadget 0x400565
Stack change: 0x8
Changed registers: rax rdi rbx 
Register dependencies:
rax rax rax rdi rdi rdi rdi rbx rax rax 
```
TODO:
Update the changed registers based on the `init_val` and `new_val`
The dependencies list(still in progress, has to be  verbose) is memory read and write done on the registers in the gadget.
Stack change should be updated.
Show jump or syscall
Fix leaks

Commit:
Nit change

Commit:
Cleanup and add memory rw support

- Code refactor and fixed some bugs related to IL event read
- memory read and var read is fixed
- Fix mem leaks

giridharprasath commented:
Current output:

```
  0x004003d0                 a6  cmpsb byte [rsi], byte ptr [rdi]
  0x004003d1               0440  add al, 0x40
  0x004003d3               00ff  add bh, bh
  0x004003d5         15160c2000  adc eax, 0x200c16
  0x004003da                 f4  hlt
Gadget size: 11
Gadget 0x4003d0
Stack change: 0x0
Changed registers: rsi rdi rax rbx 
Register dependencies:
Memory Read: rsi 5
Memory Read: rdi 4195495
Var write: rax 4195648 4195712
Var write: rbx 0 0
Var write: rbx 0 0
Var write: rax 4195648 6295958
```

Added support for stack change
Add the memory read and write to dependencies list
Track the register state in `modified_registers`

Another eg:

```
  0x0040052d                 5c  pop rsp
  0x0040052e               415d  pop r13
  0x00400530               415e  pop r14
  0x00400532               415f  pop r15
  0x00400534                 c3  ret
Gadget size: 8
Gadget 0x40052d
Stack change: 0x28
Changed registers: rsp r13 r14 r15 
Register dependencies:
```

Commit:
fix dependency function flags

Commit:
Generate accurate Rop gadgets

giridharprasath commented:
Current output:

![ScreenShot_2024-06-27_at_12:26:01-AM](https://github.com/rizinorg/rizin/assets/17312540/46c6b1e5-30bc-4603-9c9d-ad9802ab5e9e)

## TODO:
Refactor `construct_rop_gadget` API and move the il event handling code to `Rg` handler
Tidy up the `Rg` cmd output
Fix Unit and regression TC

Commit:
Regression tc fix

XVilka commented on librz/arch/analysis.c:
```diff
@@ -240,6 +240,22 @@ RZ_API char *rz_analysis_get_reg_profile(RzAnalysis *analysis) {
 		: NULL;
 }
 
+RZ_API bool rz_analysis_is_reg_in_profile(RzAnalysis *analysis, const char *str) {
```
When you add new API function, please add Doxygen for it as well

XVilka commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,74 @@
+
+#include <rz_cmd.h>
+
+typedef struct rz_reg_info_t {
```
When you create new include with the API, please document types (structures, enums) using Doxygen syntax as well.

XVilka replied:
Also I recommend a file-level note what this header for

XVilka commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,74 @@
+	bool is_pc_write;
+	bool is_syscall;
+	RzPVector /*<RzRegInfo*>*/ *modified_registers;
+	RzList /*<RzRegInfo*>*/ *dependencies;
```
I think you need spaces in those for `rz-bindgen` linter not to complain

XVilka commented on librz/core/cmd/cmd_search_rop.c:
Since you remove the ESIL code, please mark the PR as partially solving https://github.com/rizinorg/rizin/issues/2080

Commit:
Fix regression and add doc

Commit:
fix sanity

Commit:
Fix sanity switch/case

giridharprasath commented on test/db/cmd/cmd_rop:
```diff
@@ -336,19 +478,16 @@ Gadget size: 12
   0x00000142           0e946b04  call 0x8d6
-Gadget size: 12
+Gadget size: 10
 
```
Updated this TC as it is a duplicate and not printed out, hence the Gadget starts from `13c`

giridharprasath commented on test/db/cmd/cmd_rop:
```diff
@@ -336,19 +478,16 @@ Gadget size: 12
   0x0000015e               8fe1  ldi r24, 0x1f
   0x00000160               92e0  ldi r25, 0x02
   0x00000162           0e946b04  call 0x8d6
-Gadget size: 12
```
Same here

Commit:
Tiny rop gadget print fix

XVilka reviewed (changes requested):
```
Run python3 rz-bindgen/src/lint.py \
Missing type comment at <rizin/librz/core/rop.c:71:77> (token is not a comment)
Missing type comment at <rizin/librz/core/rop.c:84:77> (token is not a comment)
Missing type comment at <rizin/librz/core/rop.c:110:80> (token is not a comment)
Missing type comment at <rizin/librz/core/rop.c:149:76> (token is not a comment)
Missing type comment at <rizin/librz/core/rop.c:392:142> (token is not a comment)
Type comment at <rizin/librz/core/rop.c:509:77> for unknown type RzCoreAsmHit
Missing type comment at <rizin/librz/core/rop.c:821:102> (token is not a comment)
Missing type comment at <rizin/librz/core/cmd/cmd_search_rop.c:155:33> (token is not a comment)
Missing type comment at <rizin/librz/core/cmd/cmd_search_rop.c:382:16> (token is not a comment)
```

Commit:
fix Linter comments

giridharprasath commented:
I think we can merge this PR(once review is over) and we can improve and add direct ROP gadget search in separate PR @XVilka @wargio

wargio commented on librz/core/rop.c:
```diff
@@ -538,7 +538,7 @@ static void analyze_gadget(RzCore *core, RzCoreAsmHit /*<RzCoreAsmHit *>*/ *hit,
 
 void print_rop_gadget_info(RzCore *core, RzRopGadgetInfo *gadget_info) {
-	rz_cons_printf("Gadget 0x%" PFMT64x "\n", gadget_info->address);
+	rz_cons_printf("\nGadget 0x%" PFMT64x "\n", gadget_info->address);
```
why the additional newline ?

giridharprasath replied:
Separator between gadget output.
`\Rg` output:
```
Gadget 0x401004
Stack change: 0x8
Changed registers: rax rsp 
Register dependencies:
Var write: rax 128 140
Var write: rax 128 140

Gadget 0x401005
Stack change: 0x8
Changed registers: rax rsp 
Register dependencies:
Memory Read: rax 0
```

wargio replied:
ok, it makes sense.

wargio commented on librz/arch/analysis.c:
```diff
@@ -240,6 +240,31 @@ RZ_API char *rz_analysis_get_reg_profile(RzAnalysis *analysis) {
+		return false;
+	}
+
+	if (strstr(reg_prof, str) != NULL) {
```
```suggestion
	if (strstr(reg_prof, str)) {
```

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
-	if (!str) {
-		return NULL;
+static void skip_whitespace(const char *str, int *idx) {
+	while (str[*idx] == ' ' || str[*idx] == '\t' || str[*idx] == '\n' || str[*idx] == '\r') {
```
```suggestion
	while (IS_WHITECHAR(str[*idx])) {
```

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
+
+	skip_whitespace(str, idx);
+
+	if (str[*idx] == '0' && (str[*idx + 1] == 'x' || str[*idx + 1] == 'X')) {
```
this might be an out of bound read. @XVilka ?

XVilka replied:
Right. Also `idx` pointer should be marked as `RZ_NONNULL` and have an assert.

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
+	rop_constraint->args[DST_REG] = dst_reg;
+	rop_constraint->args[SRC_REG] = NULL;
+	char value_str[256];
+	snprintf(value_str, sizeof(value_str), "%llu", const_value);
```
```suggestion
	rz_strf(value_str, "%" PFMT64u, const_value);
```

wargio replied:
`%llu` is not portable, you need to use `"%" PFMT64u` which gets converted to the correct format.

wargio replied:
after re-reading the code since you are going to use `strdup` anyway, then use `rz_str_newf("%" PFMT64u, const_value);`. apply this to all the `snprintf(value_str, sizeof(value_str), "%llu", const_value);` which you have defined.

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
-				    (RzListComparator)strcmp, NULL)) {
-				continue;
-			}
+	RzILOpPureCode *op_ptr = malloc(sizeof(RzILOpPureCode));
```
```suggestion
	RzILOpPureCode *op_ptr = RZ_NEW0(RzILOpPureCode);
```

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
+		return false;
+	}
+	RzList *args = rz_list_new();
+	if (!parse_il_op(args, str, &idx)) {
```
```suggestion
	if (!args || !parse_il_op(args, str, &idx)) {
```

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
-	return NULL;
+
+	char op_str[16];
+	snprintf(op_str, sizeof(op_str), "%s", rz_il_op_pure_code_stringify(*op));
+	rop_constraint->args[OP] = strdup(op_str);
+	char value_str[256];
+	snprintf(value_str, sizeof(value_str), "%llu", const_value);
```
use `rz_strf` like above and convert that `%llu` like above.

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
-	FREE_ROP;
-	return NULL;
+
+	char op_str[16];
+	snprintf(op_str, sizeof(op_str), "%s", rz_il_op_pure_code_stringify(*op));
+	rop_constraint->args[OP] = strdup(op_str);
```
there is no need to copy the output of `rz_il_op_pure_code_stringify` twice. just

```c
const char *op_str = rz_il_op_pure_code_stringify(*op);
rop_constraint->args[OP] = rz_str_dup(op_str);
```

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,402 @@
 
-	rz_list_foreach (ropList, iter_r, esil_str) {
-		fillRegisterValues(core);
+	char op_str[16];
+	snprintf(op_str, sizeof(op_str), "%s", rz_il_op_pure_code_stringify(*op));
+	rop_constraint->args[OP] = strdup(op_str);
```
same here as above. this is useless. just follow the guidelines above.

wargio commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			break;
+		}
+		// reg_info_dup->init_val = rz_bv_to_ut64(init_val);
+		reg_info_dup->new_val = rz_bv_to_ut64(new_val);
```
this works, but is also kinda limited to 64 bits. @XVilka what do you think?

XVilka replied:
For now it's ok, but could be improved as more architectures will be added for working with the upcoming ROP generator.

Rot127 replied:
But please let's do the `instruction word` abstraction first. It is a pain to work with Hexagon and these kind of analysis algos.

wargio replied:
i think that step should be handled by the disassembler (RzArch) and should be defined by the arch, so we could have a packet with 1 or N instructions.

XVilka replied:
@DMaroo @thestr4ng3r do you have an opinion on this?

giridharprasath replied:
I will be opening a new issue for tracking this.

wargio replied:
i think a comment here is enough. just say that this is limited to max 64 bit val but it could be expanded in the future.

wargio commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		if (!val) {
+			break;
+		}
+		reg_info = rz_reg_info_new(core, event, rz_bv_to_ut64(val), rz_bv_to_ut64(val));
```
define a `ut64 val64 = rz_bv_to_ut64(val);` then pass `val64` as argument for `rz_reg_info_new`

wargio reviewed (changes requested):
Overall looks very well thought. just some minor changes needed for me (see comments).
That said, well done.

Rot127 commented on librz/core/cmd_descs/cmd_descs.c:
```diff
@@ -1397,10 +1398,26 @@ static const RzCmdDescArg cmd_query_gadget_args[] = {
+	},
+	{ 0 },
+};
+static const RzCmdDescHelp cmd_detail_gadget_help = {
```
Missing newline

wargio replied:
i believe this is auto-gen.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+// SPDX-FileCopyrightText: 2010-2021 z3phyr <giridh1337@gmail.com>
```
Year should be 2024

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	if (aop->family == RZ_ANALYSIS_OP_FAMILY_SECURITY) {
+		return false;
+	}
+	switch (aop->type) {
```
`default` case missing

Rot127 replied:
Also below

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	RzListIter *iter_dst;
+	RzRegItem *item_dst;
+	rz_list_foreach (head, iter_dst, item_dst) {
+		if (!strcmp(name, item_dst->name) && item_dst->type == RZ_REG_TYPE_GPR) {
```
There is a `RZ_STR_EQ` macro somewhere. Better use this one.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	void **it;
+	rz_pvector_foreach (gadget_info->modified_registers, it) {
+		reg_info = *it;
+		if (strcmp(reg_info->name, name) == 0) {
```
Also here `RZ_STR_EQ`

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	}
+}
+
+RzRegInfo *rz_reg_info_dup(RzRegInfo *src) {
```
These and other functions should be marked as `RZ_IPI` if not static.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	return reg_item->name;
+}
+
+bool is_base_pointer(const RzCore *core, const char *name) {
```
If only used here, please make them static. The name is pretty generic. Otherwise add a prefix to the name

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+static void analyze_gadget(RzCore *core, const RzCoreAsmHit *hit, RzRopGadgetInfo *rop_gadget_info) {
+	if (!core->analysis) {
+		return;
+	}
```
Shouldn't those checks be `rz_return_if_fail()`? Or is `core->analysis` being NULL a legal value? Also above.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		search_itv.size = UT64_MAX;
+	}
+
+	Sdb *gadgetSdb = NULL;
```
Did not review far enough, but is a `Sdb` really necessary here? We want to remove it soon. So using it at a place like this is not a good idea. Would be a `ht_ss` enough?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		return false;
+	}
+
+	if (!strcmp(arch, "mips")) { // MIPS has no jump-in-the-middle
```
RZ_STR_EQ also below

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		while (tok) {
+			rx = strdup(tok);
+			rz_list_append(rx_list, rx);
+			tok = strtok(NULL, ";");
```
I don't get this line, isn't `tok` always `NULL` after a single iteration?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			break;
+		}
+		delta = to - from;
+		buf = calloc(1, delta);
```
calloc -> `RZ_NEW0`

giridharprasath replied:
`RZ_NEW0` is for struct right, retaining this

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			result = false;
+			goto bad;
+		}
+		(void)rz_io_read_at(core->io, from, buf, delta);
```
Handle errors in read here.

Rot127 replied:
Specifically check the `read_at` and `read_at_mapped` (or a similar name). You want the `mapped` version here and don't search for gadgets in non-mapped regions.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		(void)rz_io_read_at(core->io, from, buf, delta);
+
+		// Find the end gadgets.
+		for (i = 0; i < delta; i += increment) {
```
```suggestion
		for (size_t i = 0; i < delta; i += increment) {
```

No need to define it above. Always has the risk that it is not reset properly.

Rot127 replied:
Can this whole loop be move to it's own `static` function? It would make the code way more readable.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			}
+			if (is_end_gadget(&end_gadget, crop)) {
+				RzRopEndListPair *epair = RZ_NEW0(RzRopEndListPair);
+				if (epair) {
```
You can do a `rz_return_if_fail(epair)` and save indentation. If `calloc` fails everything is over anyways.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		}
+		rz_list_reverse(end_list);
+		// If we have no end gadgets, just skip all of this search nonsense.
+		if (!rz_list_empty(end_list)) {
```
This whole `if` statement is quite deep (meaning many indentations). This is hard to read. I would suggest to:

- Move some code to their own small `static` functions
- Add one or two clean-up gotos, so you can do error handling like `if (somethign_invalid) { goto cleanup }`. This is one of the still valid `goto` use cases in C.

Before you refactor with `gotos` check again that actually everything is freed properly in case of error.
Then add them. Added suggestions below.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+						i = prev - max_inst_size_x86;
+					}
+				} else {
+					if (i < prev) {
```
`else if`

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+				}
+				if (i < 0) {
+					i = 0;
+				}
```
In what case is `i` smaller then 0?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	int delta = 0;
+	ut8 *buf;
+	RzIOMap *map;
+	HtSU *unique_hitlists = ht_su_new(HT_STR_DUP);
```
Quite a lot of those declarations can be made in place with definition. There is no need for C89 compatibility. And this pattern makes code harder to read and is error prone.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+						rz_asm_op_free(asmop);
+						asmop = NULL;
+						rz_strbuf_free(sb);
+						continue;
```
These things would belong to a `goto` block

Rot127 replied:
It repeats below again.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+						continue;
+					}
+					rz_strbuf_free(sb);
+					if (gadgetSdb) {
```
This check should have been handled before. It doesn't make sense that the db is undefined, but the logic went on so far, doesn't it?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+					} else {
+						end_gadget = NULL;
+						break;
+					}
```
Could this be maybe move to a `set_next_end_gadget()`?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+ * If unavailable, performs a ROP search with the input.
+ */
+RZ_API RzCmdStatus rz_core_rop_gadget_info(RzCore *core, const char *input, RzCmdStateOutput *state) {
+	Sdb *gadgetSdb = sdb_ns(core->sdb, "gadget_sdb", false);
```
Please check also here if a `ht_ss` wouldn't be enough. @XVilka How forbidden is `SDB` usage these days?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+
+	if (!gadgetSdb) {
+		rz_core_search_rop(core, input, 0, RZ_ROP_GADGET_PRINT, state);
+		return RZ_CMD_STATUS_OK;
```
Shouldn't this be an error code?

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			print_rop(core, hitlist, state);
+		}
+	}
+	if (type & RZ_ROP_GADGET_DETAIL) {
```
Negate check and return early.

giridharprasath replied:
Changed this to different logic

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		rz_list_free(hitlist);
+		ht_uu_free(localbadstart);
+		return NULL;
+	}
```
Combine these two if conditions. Saves a few lines of duplicate code.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			end_gadget_cnt++;
+		}
+		const int opsz = aop.size;
+		// opsz = rz_strbuf_length (asmop.buf);
```
```suggestion
```

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		const int opsz = aop.size;
+		// opsz = rz_strbuf_length (asmop.buf);
+		char *opst = aop.mnemonic;
+		RzAsmOp asmop;
```
```suggestion
		RzAsmOp asmop = RZ_EMPTY;
```

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		char *opst = aop.mnemonic;
+		RzAsmOp asmop;
+		int asm_ret = rz_asm_disassemble(core->rasm, &asmop, buf + idx, buflen - idx);
+		if (!opst) {
```
Because you passed `RZ_ANALYSIS_OP_MASK_DISASM` to the `rz_analysis_op` it already called `rz_asm_disassemble` and if it succeeded the `opst` is definitely set. So this test could be skipped.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		}
+
+		hit = rz_core_asm_hit_new();
+		if (hit) {
```
`rz_return_if_fail()` and remove `if (hit)`. If initialization fails, there is no need to continue.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		addr += opsz;
+		if (rx) {
+			grep_find = rz_regex_contains(rx, opst, RZ_REGEX_ZERO_TERMINATED, RZ_REGEX_EXTENDED, RZ_REGEX_DEFAULT);
+			search_hit = (end && grep && grep_find);
```
Define `search_hit` over the `if (rx)` as `bool search_hit`. It is defined as `int` above, but this gives the wrong idea what the variable does.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			if (rz_list_length(rx_list) > 0) {
+				rx = rz_list_get_n(rx_list, count++);
+			}
+		}
```
Move grep parsing to static function.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		}
+	}
+
+	bool found;
```
```suggestion
	bool found = false;
```

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	char *rx = NULL;
+	HtUUOptions opt = { 0 };
+	HtUU *localbadstart = ht_uu_new_opt(&opt);
+	int count = 0;
```
Please define variables in place.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	return true;
+}
+
+// TODO: follow unconditional jumps
```
Don't forget to open an issue about this one. Would suggest to remove the comment, and add this as a task to the PR description. After merge you can open the issue.

giridharprasath replied:
Noted :+1:

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		}
+	}
+}
+static void print_rop(const RzCore *core, const RzList /*<RzCoreAsmHit *>*/ *hitlist, RzCmdStateOutput *state) {
```
Missing newline

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+			rz_bv_free(old_val);
+			rz_bv_free(new_val);
+		}
+	} break;
```
Move the breaks within the `{}`

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+				rz_bv_to_ut64(new_val));
+			rz_rop_gadget_info_add_register(gadget_info, reg_info, is_dependency);
+			rz_bv_free(old_val);
+			rz_bv_free(new_val);
```
You can invert the `if (!old_val || !new_val)` condition and do 
```c
rz_bv_free(old_val);
rz_bv_free(new_val);
```

only once.

Rot127 replied:
Above as well.

Rot127 commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,100 @@
+	bool is_mem_write; ///< Register involved in Memory write.
+	ut64 init_val;
+	ut64 new_val;
+} RzRegInfo;
```
```suggestion
} RzRopRegInfo;
```

Otherwise it follows the naming in `rz_reg.h`.
Alos please document the last three fields.

Rot127 commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,100 @@
+	MOV_OP_CONST, ///< reg <- reg OP const
+	MOV_OP_REG, ///< reg <- reg OP reg
+	SYSCALL, ///< syscall
+} RzILInstructionType;
```
Naming is off here as well. Please use a `RzRop` prefix

Rot127 commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,100 @@
+
+// ROP Constraint APIs
+RZ_API void rz_rop_constraint_free(RZ_NULLABLE void *data);
+RZ_API RzList /*<RzRopConstraint *>*/ *rz_rop_constraint_list_new(void);
```
Naming should be also `rz_core_<something with rop>()`

Rot127 commented on test/unit/test_rop_constraint.c:
```diff
@@ -0,0 +1,142 @@
+// SPDX-FileCopyrightText: 2021 z3phyr <giridh1337@gmail.com>
```
2024

Rot127 commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,100 @@
+// Command APIs
+RZ_API int rz_core_search_rop(RzCore *core, const char *greparg, int regexp, RzRopRequestMask type, RzCmdStateOutput *state);
+RZ_API RzCmdStatus rz_core_rop_gadget_info(RzCore *core, const char *input, RzCmdStateOutput *state);
+RZ_API bool analyze_constraint(RzCore *core, char *str, RzRopConstraint *rop_constraint);
```
Prefix in name missing.

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+		}
+	} break;
+	case RZ_IL_EVENT_PC_WRITE: {
+		// RzILEventPCWrite *pc_write = &event->data.pc_write;
```
```suggestion
```

Rot127 commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1248 @@
+	return dup;
+}
+
+static bool is_stack_pointer(const RzCore *core, const char *name) {
+	RzRegItem *reg_item = rz_reg_get(core->analysis->reg, name, RZ_REG_TYPE_GPR);
+	if (!reg_item) {
+		return false;
+	}
+	if (core->analysis->bits == 32 && !strcmp(core->analysis->cpu, "x86")) {
+		return !strcmp(reg_item->name, "esp");
+	}
+	if (core->analysis->bits == 64) {
+		return !strcmp(reg_item->name, "rsp") || !strcmp(reg_item->name, "esp");
+	}
+	if (core->analysis->bits == 64 && !strcmp(core->analysis->cpu, "arm")) {
+		return !strcmp(reg_item->name, "r13");
+	}
+	return reg_item->name;
+}
+
+bool is_base_pointer(const RzCore *core, const char *name) {
+	RzRegItem *reg_item = rz_reg_get(core->analysis->reg, name, RZ_REG_TYPE_GPR);
+	if (!reg_item) {
+		return false;
+	}
+	if (core->analysis->bits == 32 && !strcmp(core->analysis->cpu, "x86")) {
+		return !strcmp(reg_item->name, "ebp");
+	}
+	if (core->analysis->bits == 64) {
+		return !strcmp(reg_item->name, "rbp") || !strcmp(reg_item->name, "ebp");
+	}
+	if (core->analysis->bits == 64 && !strcmp(core->analysis->cpu, "arm")) {
+		return !strcmp(reg_item->name, "r11");
+	}
+	return reg_item->name;
+}
```
These `is_stack/base_ptr` should query the register profile and compare the register name against the `SP`/`BP` defined there.
Otherwise this code is only valid for `x86` and `arm`.
Also, returning `reg_item->name` as `bool` is not the way we do it here :)

Rot127 replied:
Check out `rz_reg_role_by_name()` (or adjacent functions in `rz_reg.h`. These two functions become way smaller this way as well.

giridharprasath replied:
I will be opening an issue and finish that in a separate PR

Rot127 reviewed (changes requested):
In general really nice! I have a few remarks though (don't be too discouraged by the number of comments, most of them are suggestions and syntax stuff).

- My highest priority change request is: Please definitely refactor `rz_core_search_rop` and `construct_rop_gadget`. They are very long and have a lot of indentations. This is hard to read, maintain and debug. I left you some suggestions how to refactor. In general though: With so many variables in each function, I'd suggest to think about, what steps can be moved to separated `static` functions. This will help everyone enormously in the future to maintain this code and find bugs.
- Please add `RZ_OWN`/`RZ_BORROW` to at least the `API` functions. And to the longer `IPI` functions. And definitely if a parameter is freed by a function.
- Then there are a lot of code duplicates in error handling (`fee()` mostly). Please try remove them by refactoring the code. I made some suggestions.
- Besides, there is really no need to define all variables at the beginning of functions. Rizin is not a C89 project (and if it is and I just didn't see the memo, it definitely doesn't work :D ).

Awesome job though!

giridharprasath commented:
Phew :sweat_smile:  Thanks for the review, @Rot127  and @wargio . I didn't refactor `rz_core_search_rop` and `construct_rop_gadget` because I was concerned about potentially introducing bugs. I will definitely try to refactor both of these and address the other comments.

Rot127 commented:
> I was concerned about potentially introducing bugs. 

Then you should refactor as well :)
It is way easier to find bugs, if everything is brief and concise. Because less stuff to look at and less stuff to remember.

Commit:
Initial refactoring

Commit:
Final refactoring and test case

XVilka commented on test/integration/test_project_migrate.c:
```diff
@@ -1020,6 +1028,7 @@ int all_tests() {
 	mu_run_test(test_load_v15_seek_history);
 	mu_run_test(test_load_v15_str_config);
 	mu_run_test(test_load_v16);
+	mu_run_test(test_load_v17);
```
You need to add a test also for a migration itself to the v17

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1344 @@
+	free(gadget_info);
+}
+
+RZ_API void rz_core_rop_gadget_info_add_register(RzRopGadgetInfo *gadget_info, RzRopRegInfo *reg_info, const bool is_dependency) {
```
For every `RZ_API` that you add or change, except maybe obvious basic `*_free()` and `*_new()` functions, please add a Doxygen documentation. Moreover, for the external API it's inmportant to provide hints like `RZ_BORROW`, `RZ_OUT`, `RZ_NULLABLE`, `RZ_NONNULL` to the corresponding parameters, and for those marked with `RZ_NONNULL` add `rz_return_*()` asserts at the beginning of the function.

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1344 @@
+	}
+}
+
+RZ_API RzRopRegInfo *rz_core_rop_gadget_info_get_modified_register(RzRopGadgetInfo *gadget_info, const char *name) {
```
If function returns pointer - add an onwership (`RZ_OWN`/`RZ_BORROW`) as appropriate.

Commit:
Refactoring and doxygen fixes

Commit:
Migration test fix

Commit:
Fix migration intergration issues

Commit:
Linter fixes

XVilka commented:
```
librz/include/rz_project.h:64:84: error: code should be clang-formatted [-Wclang-format-violations]
RZ_API bool rz_project_migrate_v16_v17(RzProject *prj, RzSerializeResultInfo *res);
```

XVilka commented on test/prj/v15-seek-history.rzdb:
```diff
@@ -55,7 +55,7 @@ prj.file=./test/prj/v15-seek-history.rzdb
 
 /core/file
-relative=../bins/elf/ls
+relative=../bins/elf/crackme0x05
```
Old test projects shouldn't be changed

Commit:
Linter fixes

Commit:
Ignore test sync db

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1238 @@
+	return 0;
+}
+
+static int rz_rop_print_table_mode(const RzCore *core, const RzCoreAsmHit *hit, const RzList /*<RzCoreAsmHit *>*/ *hitlist,
```
This one could be converted to returning `bool` since only two values are possible

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1238 @@
+	}
+}
+
+static int rz_rop_process_asm_op(const RzCore *core, const RzCoreAsmHit *hit, RzAsmOp *asmop, RzAnalysisOp *aop, unsigned int *size, char **asmop_str, char **asmop_hex_str) {
```
Same

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1238 @@
+	return 0;
+}
+
+static int rz_rop_print_quiet_mode(const RzCore *core, const RzCoreAsmHit *hit, unsigned int *size, const bool colorize) {
```
Same

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1238 @@
+	return 0;
+}
+
+static int rz_rop_print_standard_mode(const RzCore *core, const RzCoreAsmHit *hit,
```
Same

XVilka commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1238 @@
+	return 0;
+}
+
+static int rz_rop_print_json_mode(const RzCore *core, const RzCoreAsmHit *hit, unsigned int *size, PJ *pj) {
```
Same

XVilka reviewed (changes requested):
Looks better! Only few small nitpicks left, apart from that - LGTM.

Commit:
Review fixes

XVilka reviewed (approved):
LGTM. Next time please use different branch name, by the way.

wargio commented on librz/core/cconfig.c:
```diff
@@ -3735,8 +3735,7 @@ RZ_API int rz_core_config_init(RzCore *core) {
 	SETI("rop.len", 5, "Maximum ROP gadget length");
-	SETBPREF("rop.sdb", "false", "Cache results in sdb (experimental)");
-	SETBPREF("rop.db", "true", "Categorize rop gadgets in sdb");
+	SETBPREF("rop.cache", "false", "Cache rop gadget results(experimental)");
```
```suggestion
	SETBPREF("rop.cache", "false", "Cache rop gadget results (experimental)");
```

wargio commented on librz/arch/analysis.c:
```diff
@@ -240,6 +243,31 @@ RZ_API char *rz_analysis_get_reg_profile(RzAnalysis *analysis) {
+ * This function checks if the given register name is present
+ * in the register profile of the given RzAnalysis.
+ */
+RZ_API bool rz_analysis_is_reg_in_profile(RzAnalysis *analysis, const char *str) {
+	rz_return_val_if_fail(analysis, false);
```
Add `RZ_NONNULL` to both args and change the assert to `rz_return_val_if_fail(analysis && str, false);`

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,457 @@
 #include <stddef.h>
```
Add your name as Copyright by adding a new `// SPDX-FileCopyrightText: ` on top of the old one.

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -4,726 +4,457 @@
+	rop_constraint->type = MOV_CONST;
+	rop_constraint->args[DST_REG] = dst_reg;
+	rop_constraint->args[SRC_REG] = NULL;
+	char value_str[256];
+	rz_strf(value_str, "%" PFMT64u, const_value);
+	rop_constraint->args[SRC_CONST] = strdup(value_str);
```
use `rz_str_newf` directly as `rop_constraint->args[SRC_CONST] = rz_str_newf("%" PFMT64u, const_value);`

wargio commented on librz/core/rop.c:
```diff
@@ -0,0 +1,1238 @@
+}
+
+static void set_increment_based_on_arch(const RzCore *core, const char *arch, int *increment) {
+	if (RZ_STR_EQ(arch, "mips")) { // MIPS has no jump-in-the-middle
+		*increment = 4;
+	} else if (RZ_STR_EQ(arch, "arm")) { // ARM has no jump-in-the-middle
+		*increment = rz_config_get_i(core->config, "asm.bits") == 16 ? 2 : 4;
+	} else if (RZ_STR_EQ(arch, "avr")) { // AVR is halfword aligned.
+		*increment = 2;
+	}
```
@XVilka it would be nice to have an issue for this. seems that this should be an info that RzArch should be able to provide.

wargio commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,136 @@
+#define RZ_ROP_H
+
+#include <rz_cmd.h>
+
```
Add the cpp barriers.

```c
#ifdef __cplusplus
extern "C" {
#endif
```

wargio commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,136 @@
+RZ_IPI void rz_core_rop_reg_info_free(RzRopRegInfo *reg_info);
+RZ_IPI RzRopRegInfo *rz_core_rop_reg_info_new(const RzCore *core, const RzILEvent *evt, ut64 init_val, ut64 new_val);
+RZ_BORROW RZ_API RzRopRegInfo *rz_core_rop_gadget_info_get_modified_register(const RZ_NONNULL RzRopGadgetInfo *gadget_info, RZ_NONNULL const char *name);
+
```
Add the cpp barrier.

```c
#ifdef __cplusplus
}
#endif
```

wargio reviewed (changes requested):
Huge improvements. fix the comments i left and i think we can merge.

Commit:
Review comment fixes

wargio commented on librz/include/rz_rop.h:
```diff
@@ -0,0 +1,141 @@
+#ifndef RZ_ROP_H
+#define RZ_ROP_H
+
+#include <rz_cmd.h>
+#ifdef __cplusplus
+extern "C" {
+#endif
+/**
+ * \file rz_rop.h
+ * \brief Return-Oriented Programming (ROP) related APIs and structures..
+ *
+ * This file contains definitions, structures, and function prototypes for handling ROP gadgets and constraints.
+ */
```
```suggestion
/**
 * \file rz_rop.h
 * \brief Return-Oriented Programming (ROP) related APIs and structures..
 *
 * This file contains definitions, structures, and function prototypes for handling ROP gadgets and constraints.
 */

#include <rz_cmd.h>

#ifdef __cplusplus
extern "C" {
#endif
```

wargio committed:
Moved around lines for readability

wargio commented on librz/core/cmd/cmd_search_rop.c:
```diff
@@ -1,729 +1,458 @@
-// SPDX-FileCopyrightText: 2009-2016 Alexandru Caciulescu <alex.darredevil@gmail.com>
+// SPDX-FileCopyrightText: 2024 z3phyr <giridh1337@gmail.com>
```
```suggestion
// SPDX-FileCopyrightText: 2024 z3phyr <giridh1337@gmail.com>
// SPDX-FileCopyrightText: 2009-2016 Alexandru Caciulescu <alex.darredevil@gmail.com>
```

wargio committed:
Fixed copyright in librz/core/cmd/cmd_search_rop.c

wargio commented on librz/arch/analysis.c:
```diff
@@ -240,6 +243,31 @@ RZ_API char *rz_analysis_get_reg_profile(RzAnalysis *analysis) {
+/**
+ * \brief Check if a register is in the analysis profile.
+ * \param analysis Pointer to the RzAnalysis object.
+ * \param str The register name to check.
+ * \return true if the register name is found, false otherwise.
+ *
+ * This function checks if the given register name is present
+ * in the register profile of the given RzAnalysis.
+ */
+RZ_API bool rz_analysis_is_reg_in_profile(RzAnalysis *analysis, const char *str) {
+	rz_return_val_if_fail(analysis && str, false);
+
+	char *reg_prof = rz_analysis_get_reg_profile(analysis);
+	if (!reg_prof) {
+		return false;
+	}
+
+	if (strstr(reg_prof, str)) {
+		free(reg_prof);
+		return true;
+	}
+	free(reg_prof);
+	return false;
+}
```
```suggestion
 * \param name The register name to check.
 * \return true if the register name is found, false otherwise.
 *
 * This function checks if the given register name is present
 * in the register profile of the given RzAnalysis.
 */
RZ_API bool rz_analysis_is_reg_in_profile(RZ_NONNULL RzAnalysis *analysis, RZ_NONNULL const char *name) {
	rz_return_val_if_fail(analysis && name, false);

	char *reg_prof = rz_analysis_get_reg_profile(analysis);
	if (!reg_prof) {
		return false;
	}

	if (strstr(reg_prof, name)) {
		free(reg_prof);
		return true;
	}
	free(reg_prof);
	return false;
}
```

wargio committed:
Rename var in function librz/arch/analysis.c

wargio commented on librz/include/rz_analysis.h:
```diff
@@ -1573,6 +1574,7 @@ RZ_API int rz_analysis_archinfo(RzAnalysis *analysis, RzAnalysisInfoType query);
 RZ_API bool rz_analysis_use(RzAnalysis *analysis, const char *name);
 RZ_API bool rz_analysis_set_reg_profile(RzAnalysis *analysis);
 RZ_API char *rz_analysis_get_reg_profile(RzAnalysis *analysis);
+RZ_API bool rz_analysis_is_reg_in_profile(RzAnalysis *analysis, const char *str);
```
```suggestion
RZ_API bool rz_analysis_is_reg_in_profile(RZ_NONNULL RzAnalysis *analysis, RZ_NONNULL const char *name);
```

wargio committed:
Fix definition in librz/include/rz_analysis.h
